### PR TITLE
feat(achievements): master buster

### LINKED ISSUE #249: Implement "Master Buster" achievement
https://docs.google.com/document/d/1aUv7YEfJOhPJ4ZctJgLZ2qf0MnoDbGGTtPWEuVyT-gc/edit
* [x] Bust 100 bots over multiple matches
* [x] 9 Point
* [x] 1 time only

### PR DESCRIPTION
**Pull request by Mautjee**
## Description

This PR introduces master buster achievement. When a player has busted 100 bots they get the achievement.
 
## Related Issues

closes #249 

## Changes Made

- [BE] Added logic for master buster achievement
- [BE] Added column for bots busted count

### CODE REVIEW

File: src/constants/main.ts [hunk 1]
```diff
@@ -45,3 +48,13 @@ export const BOTBUSTERS_LOGO_ANIMATION_SEGMENT = [
   [0, 96],
   [47, 96],
 ];
+
+export const ACHIEVEMENTS_TO_STORE: AchievementId[] = [
```
Comment (ramidecodes) [inline src/constants/main.ts:52]: I wonder if at this point we should just create achievements as records, where we store the points and "isStorable" kind of boolean.

File: src/server/service/match.ts [hunk 2]
```diff
@@ -318,13 +349,28 @@ export class Match {
...
+        await tx
           .update(users)
           .set({
             score: sql`${users.score} + ${player.score}`,
+            botsBusted: sql`${users.botsBusted} + ${player.botsBusted}`,
           })
           .where(eq(users.id, player.userId));
+
+        const playerAchievements = player.achievements
+          .filter((achievement) => {
```
Comment (ramidecodes) [inline src/server/service/match.ts:361]: for this arrow functions that immediately return an object or boolean you can declare the like:
```TS
(yourInput) => (yourReturnObjectOrBoolean)
```

No need to open up brackets

File: src/utils/date.ts [hunk 3]
```diff
@@ -78,6 +79,10 @@ export const getRelativeDate = (daysAgo: number) => {
   return formattedDate(date);
 };
 
+export const getRelativeTimeStamp = (daysAgo: number) => {
+  return Date.now() - daysAgo * 24 * 60 * 60 * 1000;
```
Comment (ramidecodes) [inline src/utils/date.ts:83]: Same as the other PR, let's add a utility for the time multipliers.

**Review by ramidecodes — APPROVED** [review]

**Review by carlos-kryha — APPROVED** [review]

File: migrations/0000_next_puck.sql [hunk 4]
```diff
@@ -1,6 +1,8 @@
 CREATE TABLE IF NOT EXISTS "bot_busters_match" (
 	"id" uuid PRIMARY KEY NOT NULL,
-	"room" json NOT NULL
+	"created_at" timestamp,
```
Comment (silimarius) [inline migrations/0000_next_puck.sql:3]: Wait... why did the migrations get modified? Did you delete any old migration file?

File: src/server/service/achievements.ts [hunk 5]
```diff
@@ -17,7 +17,7 @@ const goodBustAchievement: Achievement = {
 
 const doubleAgentAchievement: Achievement = {
   name: "Double Agent",
-  description: "Convince 2 or more humans that you are a bot in a match",
+  description: "Convince >1 human(s) that you are a bot in a match",
```
Comment (silimarius) [inline src/server/service/achievements.ts:20]:
```suggestion
  description: "Convince at least a human that you are a bot in a match",
```

Comment (silimarius) [reply src/server/service/achievements.ts:20]: Or is it the official copy?

File: src/server/service/achievements.ts [hunk 6]
```diff
@@ -28,7 +28,7 @@ const doubleAgentAchievement: Achievement = {
 
 const busterStreakAchievement: Achievement = {
   name: "Buster Streak",
-  description: "Bust three bots in a row",
+  description: "Successfully bust all bots in (3) consecutive games",
```
Comment (silimarius) [inline src/server/service/achievements.ts:31]:
```suggestion
  description: "Successfully bust all bots in 3 consecutive games",
```
unless it's official copy

File: src/server/service/achievements.ts [hunk 7]
```diff
@@ -106,13 +106,24 @@ const beginnersLuckAchievement: Achievement = {
 
 const realHumanAchievement: Achievement = {
   name: "Real Human",
-  description: "First time played as a verified human",
+  description: "Play first match as a verified human",
```
Comment (silimarius) [inline src/server/service/achievements.ts:109]:
```suggestion
  description: "Play your first match as a verified human",
```

File: src/server/service/achievements.ts [hunk 8]
```diff
@@ -106,13 +106,24 @@ const beginnersLuckAchievement: Achievement = {
...
   calculate: ({ playerAchievements }) => {
-    if (!playerAchievements) return false;
     return !alreadyReceivedAchievement(playerAchievements, "realHuman");
   },
 };
 
+const masterBusterAchievement: Achievement = {
+  name: "Master Buster",
+  description: "Busted (100) bots over multiple matches",
```
Comment (silimarius) [inline src/server/service/achievements.ts:117]:
```suggestion
  description: "Busted 100 bots over multiple matches",
```

File: src/server/service/achievements.ts [hunk 9]
```diff
@@ -106,13 +106,24 @@ const beginnersLuckAchievement: Achievement = {
...
+  calculate: ({ player, playerAchievements }) => {
+    if (alreadyReceivedAchievement(playerAchievements, "masterBuster"))
+      return false;
+    console.log(player.totalBotsBusted);
```
Comment (silimarius) [inline src/server/service/achievements.ts:121]: delete log

File: src/server/service/match.ts [hunk 10]
```diff
@@ -107,6 +107,13 @@ export class Match {
 
     this._players = lodash.shuffle([...botPlayers, ...humanPlayers]);
 
+    this.getPlayerStats().catch((err) => {
+      console.error("Error loading player history:", err);
+    });
+
+    this.checkVerifiedPlayers().catch((err) => {
+      console.error("Error checking if player is verified: ", err);
+    });
```
Comment (silimarius) [inline src/server/service/match.ts:116]: again........ this already happens in the `initMatch` function

File: src/server/service/match.ts [hunk 11]
```diff
@@ -254,6 +262,14 @@ export class Match {
           this._playerAchievements.get(player.userId)
         )
           return;
+
+        // getting total amount of bots busted by the player
+        const bots = await selectUserById(player.userId);
```
Comment (silimarius) [inline src/server/service/match.ts:267]: You are not retrieving bots, thus the variable name is misleading

**Review by silimarius — CHANGES_REQUESTED** [review]

File: src/server/service/achievements.ts (hunk 5 above)
Comment (Mautjee) [reply src/server/service/achievements.ts:20]: I took these from the official document

File: migrations/0000_next_puck.sql (hunk 4 above)
Comment (Mautjee) [reply migrations/0000_next_puck.sql:3]: The migrations are modified because there are some changes to the db

**Review by silimarius — APPROVED** [review]

**Pull request merged**